### PR TITLE
refactor: Extract DataFrame execution to class

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: bigframes/session/executor.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+# Max complexity that should be executed as a single query
+QUERY_COMPLEXITY_LIMIT = 1e7
+# Number of times to factor out subqueries before giving up.
+MAX_SUBTREE_FACTORINGS = 5
+
+_MAX_CLUSTER_COLUMNS = 4
+
+
+class BigQueryCachingExecutor:
```
Comment (chelsea-lin) [inline bigframes/session/executor.py:46]: Could you please add some comments to briefly explain how the cached executions works.

File: bigframes/session/executor.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        self.compiler: bigframes.core.compile.SQLCompiler = (
+            bigframes.core.compile.SQLCompiler(strict=strictly_ordered)
+        )
+        self.strictly_ordered: bool = strictly_ordered
+        self._cached_executions: weakref.WeakKeyDictionary[
+            nodes.BigFrameNode, nodes.BigFrameNode
+        ] = weakref.WeakKeyDictionary()
+        self.metrics = metrics
+
+    def to_sql(
```
Comment (chelsea-lin) [inline bigframes/session/executor.py:70]: Could you please explain the functionality of the public methods `to_sql`, `execute`, and `export` in more detail? The name export isn't immediately clear to me, so a more comprehensive explanation would be helpful.

File: bigframes/session/executor.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+    def _run_execute_query(
+        self,
+        sql: str,
+        job_config: Optional[bq_job.QueryJobConfig] = None,
+        api_name: Optional[str] = None,
+    ) -> Tuple[bigquery.table.RowIterator, bigquery.QueryJob]:
+        """
+        Starts BigQuery query job and waits for results.
+        """
+        job_config = bigquery.QueryJobConfig()
```
Comment (chelsea-lin) [inline bigframes/session/executor.py:172]: The `job_config` overwrites the one in the function parameter. Could you double check if it's in purpose.

File: bigframes/session/_io/bigquery/__init__.py [hunk 4]
```diff
@@ -222,17 +223,17 @@ def add_labels(job_config, api_name: Optional[str] = None):
 
 
 def start_query_with_client(
```
Comment (chelsea-lin) [inline bigframes/session/_io/bigquery/__init__.py:225]: Can we make the `Session` as a class and then `bq_client` and `metrics` can be part of it?

**Review by chelsea-lin — CHANGES_REQUESTED** [review]

File: bigframes/session/_io/bigquery/__init__.py (hunk 4 above)
Comment (TrevorBergeron) [reply bigframes/session/_io/bigquery/__init__.py:225]: Would rather pass more precise context, this makes it more flexible for use in other places, and reduces circular dependency between files.

File: bigframes/session/executor.py (hunk 3 above)
Comment (TrevorBergeron) [reply bigframes/session/executor.py]: yeah, we don't want that. fixed in new revision

File: bigframes/session/executor.py (hunk 2 above)
Comment (TrevorBergeron) [reply bigframes/session/executor.py:73]: Added a little docstring to each

File: bigframes/session/executor.py (hunk 1 above)
Comment (TrevorBergeron) [reply bigframes/session/executor.py:46]: Added a bit of an explantation to class description.

**Review by chelsea-lin — APPROVED** [review]
Thanks for improving code readability consistently.

**Pull request merged**